src/apps/wall/l7fw.lua

```diff
@@ -0,0 +1,275 @@
+end
+
+-- called by pfmatch handler, handle rejection response
+function L7Fw:reject(pkt, len)
+   link.transmit(assert(self.output.reject,
+                        "output port for reject policy not found"),
+                 self:make_reject_response())
+   packet.free(self.current_packet)
+end
+
+-- called by pfmatch handler, forward packet
+function L7Fw:accept(pkt, len)
+   link.transmit(self.output.output, self.current_packet)
+end
+
+function L7Fw:push()
+   local i       = assert(self.input.input, "input port not found")
+   local o       = assert(self.output.output, "output port not found")
+   local rules   = self.rules
+   local scanner = self.scanner
+
+   while not link.empty(i) do
+      local pkt  = link.receive(i)
+      local flow = scanner:get_flow(pkt)
+
+      -- so that pfmatch handler methods can access the original packet
+      self.current_packet = pkt
+
+      if flow then
+         local name   = scanner:protocol_name(flow.protocol)
+         local policy = rules[name] or rules["default"]
+
+         if policy == "accept" then
+            self:accept(pkt.data, pkt.length)
+         elseif policy == "drop" then
+            self:drop(pkt.data, pkt.length)
+         elseif policy == "reject" then
+            self:reject(pkt.data, pkt.length)
+         -- handle a pfmatch string case
+         elseif type(policy) == "string" then
+            if self.handler_map[policy] then
+               -- we've already compiled a matcher for this policy
+               self.handler_map[policy](self, pkt.data, pkt.length, flow.packets)
+            else
+               local opts    = { extra_args = { "flow_count" } }
+               local handler = match.compile(policy, opts)
+               self.handler_map[policy] = handler
+               handler(self, pkt.data, pkt.length, flow.packets)
+            end
+         -- TODO: what should the default policy be if there is none specified?
+         else
+            self:accept(pkt.data, pkt.length)
+         end
+      else
+         -- TODO: we may wish to have a default policy for packets
+         --       without detected flows instead of just forwarding
+         link.transmit(o, pkt)
+      end
+   end
+end
+
+-- create either an ICMP port unreachable packet or a TCP RST to
+-- send in case of a reject policy
+function L7Fw:make_reject_response()
+   local pkt        = self.current_packet
+   local ether_orig = ether:new_from_mem(pkt.data, pkt.length)
+   local ip_orig
+
+   if ether_orig:type() == 0x0800 then
+      ip_orig = ipv4:new_from_mem(pkt.data + ether_orig:sizeof(),
+                                  pkt.length - ether_orig:sizeof())
+   elseif ether_orig:type() == 0x86dd then
+      ip_orig = ipv6:new_from_mem(pkt.data + ether_orig:sizeof(),
+                                  pkt.length - ether_orig:sizeof())
+   else
+      -- no responses to non-IP packes
+      return
+   end
+
+   local is_tcp  = false
+   local ip_protocol
+
+   if ip_orig:version() == 4 then
```

Comment:
Side note: Many of the constants from `scanner.lua` are likely to be similar or the same to the ones in `src/apps/lwaftr/constants.lua` — later on, when working on upstreaming SnabbWall, IMHO it would make sense to try and merge both.  There are constants readily available at `src/apps/wall/scanner.lua`that maybe it's reasonable to reuse. Best thing is probably to move them into `src/apps/wall/constants.lua` and `require()` that from both the `scanner.lua` and `l7fw.lua`.

---

Side note: Many of the constants from `scanner.lua` are likely to be similar or the same to the ones in `src/apps/lwaftr/constants.lua` — later on, when working on upstreaming SnabbWall, IMHO it would make sense to try and merge both.